gadget/update.go

```diff
@@ -1174,74 +1181,160 @@ func volumeStructureToLocationMapImpl(old GadgetData, mod Model, laidOutVols map
-	}
+	// collect the updates and validate that they are doable from an abstract
+	// sense first
 
-	// layout old partially, without going deep into the layout of structure
-	// content
-	pOld, err := LayoutVolumePartially(oldVol, DefaultConstraints)
-	if err != nil {
-		return fmt.Errorf("cannot lay out the old volume: %v", err)
-	}
+	// note that this code is written such that before we perform any update, we
+	// validate that all updates are valid and that all volumes are compatible
+	// between the old and the new state, this is to prevent applying valid
+	// updates on one volume when another volume is invalid, if that's the case
+	// we treat the whole gadget as invalid and return an error blocking the
+	// refresh
 
-	// Layout new volume, delay resolving of filesystem content
-	constraints := DefaultConstraints
-	constraints.SkipResolveContent = true
-	pNew, err := LayoutVolume(new.RootDir, new.KernelRootDir, newVol, constraints)
-	if err != nil {
-		return fmt.Errorf("cannot lay out the new volume: %v", err)
-	}
+	// TODO: should we handle the updates on multiple volumes in a
+	// deterministic order? iterating over maps is not deterministic, but we
+	// perform all updates at the end together in one call
+	allUpdates := []updatePair{}
+	laidOutVols := map[string]*LaidOutVolume{}
+	for volName, oldVol := range old.Info.Volumes {
+		newVol := new.Info.Volumes[volName]
 
-	if err := canUpdateVolume(pOld, pNew); err != nil {
-		return fmt.Errorf("cannot apply update to volume: %v", err)
-	}
+		if oldVol.Schema == "" || newVol.Schema == "" {
+			return fmt.Errorf("internal error: unset volume schemas: old: %q new: %q", oldVol.Schema, newVol.Schema)
+		}
 
-	if updatePolicy == nil {
-		updatePolicy = defaultPolicy
-	}
+		// layout old partially, without going deep into the layout of structure
+		// content
+		pOld, err := LayoutVolumePartially(oldVol, DefaultConstraints)
+		if err != nil {
+			return fmt.Errorf("cannot lay out the old volume %s: %v", volName, err)
+		}
 
-	// ensure all required kernel assets are found in the gadget
-	kernelInfo, err := kernel.ReadInfo(new.KernelRootDir)
-	if err != nil {
-		return err
+		// Layout new volume, delay resolving of filesystem content
+		constraints := DefaultConstraints
+		constraints.SkipResolveContent = true
+		pNew, err := LayoutVolume(new.RootDir, new.KernelRootDir, newVol, constraints)
+		if err != nil {
+			return fmt.Errorf("cannot lay out the new volume %s: %v", volName, err)
+		}
+
+		laidOutVols[volName] = pNew
+
+		if err := canUpdateVolume(pOld, pNew); err != nil {
+			return fmt.Errorf("cannot apply update to volume %s: %v", volName, err)
+		}
+
+		// ensure all required kernel assets are found in the gadget
+		kernelInfo, err := kernel.ReadInfo(new.KernelRootDir)
+		if err != nil {
+			return err
+		}
+
+		if err := gadgetVolumeConsumesOneKernelUpdateAsset(pNew.Volume, kernelInfo); err != nil {
```

Comment:
I dobule checked, this actually doesn't look correct, this check expects the volume to consume at least one kernel assets. But in a multi-volume scenario we need only one volume to do that, not all volumes to consumes at least one asset. We could argue whether the boot volumes should always consume one asset or something....

Reply:
Yes, I adjusted this now so that we just enforce that if there are kernel assets which require an update that at least one is consumed across any of the volumes